### PR TITLE
Invocation overloaded method through delegate.

### PR DESCRIPTION
**Pull request by iskiselev**
Here is fix for #806 and test for it.

### CODE REVIEW

**Comment by kg** [comment]
```Function.bind``` can't ever be used, for performance reasons. It's real bad. Global deoptimization town. =[

**Comment by iskiselev** [comment]
It should hit only delegate creation. Can you suggest how we can change it?

**Comment by kg** [comment]
If you can, take a look at https://github.com/sq/JSIL/commit/b7df487286d1b5bdf26a597f953534b40371ecf4 , since I solved a similar problem there (and got a huge performance win).

Hand-written closures and CreateNamedFunction are two reasonable choices. They don't seem like they should be faster than bind, but they are. (Though CreateNamedFunction is only faster if you cache the creation function)

**Comment by iskiselev** [comment]
Right now I don't understand how it can help me, but I'll look on it more a little bit later.
In this PR I changed delegate creation from (example):
```
$T02().New(
  $thisType, 
  $thisType.A,
  function () { return JSIL.GetMethodInfo($thisType, "A", $S00(), true); })
```
to:
```
$T02().New(
  $thisType,
  function () { return $thisType.A(); }.bind(this), 
  function () { return JSIL.GetMethodInfo($thisType, "A", $S00(), true); }.bind(this))
```
So, JSIL writes full method invocation inside function (and will use CallVirtual/CallStatic if overload resolution involved).

I suppose it also should fix resolving of MethodInfo in some cases where Generic Parameters are involved (not sure, I've not seen any bug here, but they may presented).

What exactly have performance problem - calling `bind` or calling created function after it?

**Comment by kg** [comment]
The return value of bind is a special class of function. Any time you call a function of that class, the call site is permanently deoptimized for all future calls. (This is a generalization)

**Comment by iskiselev** [comment]
Sorry, still don't understand. Which function is deoptimized:
- function on which bind was called (1)
- function which called bind (2)
- function which call result of bind (3)

(1) is really small call - it will have only function invocation (possible through signature), but called every time we call delegate
(2) in our case is function that created delegate, deoptimization here may be critical
(3) here it is `function Delegate_Invoke () {return method.apply(object, arguments); };`.

I suppose that in (1) and (3) case deoptimization should not be critical, am I wrong?

**Comment by kg** [comment]
I believe both 2 and 3 can be deoptimized. I'm not entirely certain, though. If delegate_Invoke gets deopted that's probably okay, but it could turn into an optimization boundary and prevent inlining, which can be nasty.

**Comment by iskiselev** [comment]
I don't suppose that (2), but we can change it to:
```
$T02().New(
  $thisType,
  (function(thisObject) { function () { return $thisType.A(); }.bind(thisObject)};)(this), 
  (function(thisObject) { function () { return JSIL.GetMethodInfo($thisType, "A", $S00(), true); }.bind(thisObject)};)(this))
```
It will protect (2), but I'm really unsure if we should do it.

**Comment by iskiselev** [comment]
Maybe we should write some test to compare. I'm only unsure how complex should be (2) so that we see any difference from optimization.

File: JSIL/SpecialIdentifiers.cs [hunk 1]
```diff
@@ -111,18 +111,43 @@ JSSpecialIdentifiers js
...
                     new JSType(delegateType),
-                    new JSFakeMethod("New", delegateType, new[] { TypeSystem.Object, TypeSystem.Object }, MethodTypes)
-                ), 
-                method == null ? new [] { thisReference, targetMethod } : new [] {
-                    thisReference, 
-                    targetMethod, 
-                    new JSDefferedExpression(new JSMethodOfExpression(method.Reference, method.Method, method.MethodTypes, method.GenericArguments)),  },
+                    new JSFakeMethod("New", delegateType, new[] {TypeSystem.Object, TypeSystem.Object}, MethodTypes)
+                    ),
+                jsMethod == null
```
Comment (kg) [inline JSIL/SpecialIdentifiers.cs]: Can you split this out into some named locals? This is way too complicated for a single statement.

Comment (iskiselev) [inline JSIL/SpecialIdentifiers.cs]: Done

**Comment by iskiselev** [comment]
Should I write test of performance hit on bind method in this case or @kg, you will check it yourself? Maybe you have better idea about fixing target issue?

**Comment by kg** [comment]
I'm probably going to revisit this at some point and figure out whether it needs to be changed.

**Comment by iskiselev** [comment]
Rebased to current master.

**Comment by kg** [comment]
I'll merge this and we can sort out bind later, I guess.

**Comment by iskiselev** [comment]
Should we create separate open issue to look on it?

**Comment by iskiselev** [comment]
I've broken #880 with this PR.
Problem is that JSDeferredExpression capture all variables of parent function.
Really we need walk through InnerExpression, get all referenced variables and pass them inside some additional function that will generate inner function.
We also can try to change all inner expressions so that they use self instead this if we'll go this way.

**Comment by iskiselev** [comment]
I have prototype of proper fix for both this and #875.
Will prepare PR in few days.

**Pull request merged**